File: src/engine/environment/atmosphere/state.ts
import { z } from "zod";
import type { CompiledAtmosphere, CompiledVolume } from "./definition.ts";
import type {
  AtmosphereFacts,
  AtmosphereParcel,
  AtmosphereState,
} from "./types.ts";

const finite = z.number().finite();
const parcelSchema = z.strictObject({
  volumeId: z.string().min(1).max(160),
  carrierKg: finite.nonnegative(),
  smokeKg: finite.nonnegative(),
  heatJ: finite,
});
const stateSchema = z.strictObject({
  version: z.literal("connected-atmosphere-state-v1"),
  identity: z.string().min(1),
  parcels: z.array(parcelSchema).min(1),
  initialCarrierKg: finite.positive(),
  initialSmokeKg: finite.nonnegative(),
  initialHeatJ: finite,
  smokeSourceKg: finite.nonnegative(),
  heatSourceJ: finite,
  carrierBoundaryKg: finite,
  smokeBoundaryKg: finite,
  heatBoundaryJ: finite,
});

const sum = (values: readonly number[]) =>
  values.reduce((total, value) => total + value, 0);

function tolerance(values: readonly number[]) {
  return (
    256 *
    Number.EPSILON *
    Math.max(1, ...values.map((value) => Math.abs(value))) *
    Math.max(1, values.length)
  );
}

export function parcelTemperature(
  g: CompiledAtmosphere,
  parcel: AtmosphereParcel,
) {
  if (parcel.carrierKg === 0)
    return parcel.smokeKg === 0 && parcel.heatJ === 0
      ? g.definition.ambient.temperatureK
      : Number.POSITIVE_INFINITY;
  return (
    g.definition.ambient.temperatureK +
    parcel.heatJ / (parcel.carrierKg * g.definition.model.heatCapacityJKgK)
  );
}

export function parcelPressure(
  g: CompiledAtmosphere,
  volume: CompiledVolume,
  parcel: AtmosphereParcel,
) {
  return (
    (parcel.carrierKg *
      g.definition.model.specificGasConstantJKgK *
      parcelTemperature(g, parcel)) /
    volume.volumeM3
  );
}

function validatePhysicalParcel(
  g: CompiledAtmosphere,
  volume: CompiledVolume,
  parcel: AtmosphereParcel,
) {
  const temperatureK = parcelTemperature(g, parcel),
    pressurePa = parcelPressure(g, volume, parcel),
    pressureRatio = pressurePa / g.definition.ambient.pressurePa,
    model = g.definition.model;
  if (
    !Number.isFinite(temperatureK) ||
    Math.abs(temperatureK - g.definition.ambient.temperatureK) >
      model.maxTemperatureDeltaK ||
    !Number.isFinite(pressureRatio) ||
    (parcel.carrierKg > 0 && pressureRatio < 1 / model.maxPressureRatio) ||
    pressureRatio > model.maxPressureRatio ||
    (parcel.carrierKg === 0
      ? parcel.smokeKg !== 0 || parcel.heatJ !== 0
      : parcel.smokeKg / parcel.carrierKg > model.maxSmokeMassFraction)
  )
    throw new Error(
      `atmosphere parcel ${parcel.volumeId} exceeds its envelope`,
    );
}

function balance(state: AtmosphereState) {
  const carrier = [
      ...state.parcels.map((entry) => entry.carrierKg),
      state.carrierBoundaryKg,
      -state.initialCarrierKg,
    ],
    smoke = [
      ...state.parcels.map((entry) => entry.smokeKg),
      state.smokeBoundaryKg,
      -state.smokeSourceKg,
      -state.initialSmokeKg,
    ],
    heat = [
      ...state.parcels.map((entry) => entry.heatJ),
      state.heatBoundaryJ,
      -state.heatSourceJ,
      -state.initialHeatJ,
    ];
  const result = {
    carrierKg: sum(carrier),
    smokeKg: sum(smoke),
    heatJ: sum(heat),
  };
  if (
    Math.abs(result.carrierKg) > tolerance(carrier) ||
    Math.abs(result.smokeKg) > tolerance(smoke) ||
    Math.abs(result.heatJ) > tolerance(heat)
  )
    throw new Error("atmosphere stock and ledger balance disagree");
  return result;
}

export function validateState(
  g: CompiledAtmosphere,
  input: unknown,
): AtmosphereState {
  const state = stateSchema.parse(input);
  if (
    state.identity !== g.identity ||
    state.parcels.length !== g.volumes.length ||
    state.parcels.some((entry, index) => entry.volumeId !== g.volumes[index].id)
  )
    throw new Error("atmosphere state and definition identity disagree");
  state.parcels.forEach((entry, index) =>
    validatePhysicalParcel(g, g.volumes[index], entry),
  );
  balance(state);
  return state;
}

export function copyState(state: AtmosphereState): AtmosphereState {
  return Object.freeze({
    ...state,
    parcels: Object.freeze(
      state.parcels.map((entry) => Object.freeze({ ...entry })),
    ),
  });
}

export function initialState(g: CompiledAtmosphere, input: unknown) {
  const parsed = z.array(parcelSchema).parse(input);
  if (
    parsed.length !== g.volumes.length ||
    new Set(parsed.map((entry) => entry.volumeId)).size !== parsed.length
  )
    throw new Error("atmosphere initial state needs one parcel per volume");
  const byId = new Map(parsed.map((entry) => [entry.volumeId, entry]));
  const parcels = g.volumes.map((entry) => {
    const parcel = byId.get(entry.id);
    if (!parcel)
      throw new Error(`atmosphere initial parcel missing ${entry.id}`);
    return parcel;
  });
  const state = {
    version: "connected-atmosphere-state-v1" as const,
    identity: g.identity,
    parcels,
    initialCarrierKg: sum(parcels.map((entry) => entry.carrierKg)),
    initialSmokeKg: sum(parcels.map((entry) => entry.smokeKg)),
    initialHeatJ: sum(parcels.map((entry) => entry.heatJ)),
    smokeSourceKg: 0,
    heatSourceJ: 0,
    carrierBoundaryKg: 0,
    smokeBoundaryKg: 0,
    heatBoundaryJ: 0,
  };
  validateState(g, state);
  return copyState(state);
}

export function atmosphereFacts(g: CompiledAtmosphere, input: unknown) {
  const state = validateState(g, input),
    result: AtmosphereFacts = {
      volumes: Object.freeze(
        state.parcels.map((parcel, index) => {
          const volume = g.volumes[index],
            temperatureK = parcelTemperature(g, parcel);
          return Object.freeze({
            ...parcel,
            volumeM3: volume.volumeM3,
            elevationM: volume.elevationM,
            pressurePa: parcelPressure(g, volume, parcel),
            temperatureK,
            smokeKgM3: parcel.smokeKg / volume.volumeM3,
          });
        }),
      ),
      balance: Object.freeze(balance(state)),
    };
  return Object.freeze(result);
}

File: engine/kernel/src/atmosphere/state.rs
use super::*;

impl CompiledAtmosphere {
    pub(super) fn envelope_valid(&self, index: usize, parcel: &AtmosphereParcel) -> bool {
        let temperature = self.temperature(index, parcel);
        let pressure = self.pressure(index, parcel);
        temperature.is_finite()
            && temperature > 0.0
            && (temperature - self.definition.ambient.temperature_k).abs()
                <= self.definition.model.max_temperature_delta_k
            && pressure.is_finite()
            && pressure >= 0.0
            && pressure
                <= self.definition.ambient.pressure_pa * self.definition.model.max_pressure_ratio
            && parcel.smoke_kg <= parcel.carrier_kg * self.definition.model.max_smoke_mass_fraction
    }

    pub fn initial(&self) -> AtmosphereState {
        let parcels = self
            .definition
            .volumes
            .iter()
            .map(|volume| {
                let volume_m3 = self.volume_m3[self.volume_index[&volume.id]];
                AtmosphereParcel {
                    volume_id: volume.id.clone(),
                    carrier_kg: self.ambient_carrier_density * volume_m3,
                    smoke_kg: 0.0,
                    heat_j: 0.0,
                }
            })
            .collect::<Vec<_>>();
        let initial_carrier_kg = parcels.iter().map(|parcel| parcel.carrier_kg).sum();
        AtmosphereState {
            exchange_cache: None,
            owner: self.owner.clone(),
            version: "connected-atmosphere-state-v1".into(),
            identity: self.identity.clone(),
            parcels,
            initial_carrier_kg,
            initial_smoke_kg: 0.0,
            initial_heat_j: 0.0,
            smoke_source_kg: 0.0,
            heat_source_j: 0.0,
            carrier_boundary_kg: 0.0,
            smoke_boundary_kg: 0.0,
            heat_boundary_j: 0.0,
        }
    }

    pub(super) fn validate_state(&self, state: &AtmosphereState) -> Result<(), String> {
        if !Arc::ptr_eq(&state.owner, &self.owner)
            || state.version != "connected-atmosphere-state-v1"
            || state.identity != self.identity
            || state.parcels.len() != self.definition.volumes.len()
        {
            return Err("atmosphere state binding mismatch".into());
        }
        for (index, parcel) in state.parcels.iter().enumerate() {
            if parcel.volume_id != self.definition.volumes[index].id
                || !parcel.carrier_kg.is_finite()
                || parcel.carrier_kg < 0.0
                || !parcel.smoke_kg.is_finite()
                || parcel.smoke_kg < 0.0
                || !parcel.heat_j.is_finite()
            {
                return Err("invalid atmosphere parcel".into());
            }
            if !self.envelope_valid(index, parcel) {
                return Err("atmosphere parcel exceeds physical envelope".into());
            }
        }
        for value in [
            state.initial_carrier_kg,
            state.initial_smoke_kg,
            state.initial_heat_j,
            state.smoke_source_kg,
            state.heat_source_j,
            state.carrier_boundary_kg,
            state.smoke_boundary_kg,
            state.heat_boundary_j,
        ] {
            if !value.is_finite() {
                return Err("invalid atmosphere ledger".into());
            }
        }
        let carrier = state
            .parcels
            .iter()
            .map(|parcel| parcel.carrier_kg)
            .sum::<f64>()
            + state.carrier_boundary_kg;
        let smoke = state
            .parcels
            .iter()
            .map(|parcel| parcel.smoke_kg)
            .sum::<f64>()
            + state.smoke_boundary_kg;
        let heat = state
            .parcels
            .iter()
            .map(|parcel| parcel.heat_j)
            .sum::<f64>()
            + state.heat_boundary_j;
        if (carrier - state.initial_carrier_kg).abs() > 1e-8 * state.initial_carrier_kg.max(1.0)
            || (smoke - state.initial_smoke_kg - state.smoke_source_kg).abs()
                > 1e-8
                    * (state.initial_smoke_kg + state.smoke_source_kg)
                        .abs()
                        .max(1.0)
            || (heat - state.initial_heat_j - state.heat_source_j).abs()
                > 1e-8 * (state.initial_heat_j + state.heat_source_j).abs().max(1.0)
        {
            return Err("atmosphere ledger is not conserved".into());
        }
        Ok(())
    }

    pub fn encode_state(&self, state: &AtmosphereState) -> Result<Vec<u8>, String> {
        self.validate_state(state)?;
        let bytes = postcard::to_allocvec(&(STATE_VERSION, self.content_digest, state))
            .map_err(|_| "atmosphere state encoding failed")?;
        if bytes.len() > MAX_STATE_BYTES {
            return Err("atmosphere state exceeds byte bound".into());
        }
        Ok(bytes)
    }

    pub fn decode_state(&self, bytes: &[u8]) -> Result<AtmosphereState, String> {
        if bytes.len() > MAX_STATE_BYTES {
            return Err("atmosphere state exceeds byte bound".into());
        }
        let ((version, digest, mut state), rest): ((u16, [u8; 32], AtmosphereState), &[u8]) =
            postcard::take_from_bytes(bytes).map_err(|_| "invalid atmosphere state")?;
        if !rest.is_empty() {
            return Err("atmosphere state contains trailing bytes".into());
        }
        if version != STATE_VERSION {
            return Err("unsupported atmosphere state version".into());
        }
        if digest != self.content_digest {
            return Err("atmosphere state definition binding mismatch".into());
        }
        state.owner = self.owner.clone();
        self.validate_state(&state)?;
        Ok(state)
    }
}

// Derived execution caches do not change a physical state or its wire format.
impl PartialEq for AtmosphereState {
    fn eq(&self, other: &Self) -> bool {
        self.owner == other.owner
            && self.version == other.version
            && self.identity == other.identity
            && self.parcels == other.parcels
            && self.initial_carrier_kg == other.initial_carrier_kg
            && self.initial_smoke_kg == other.initial_smoke_kg
            && self.initial_heat_j == other.initial_heat_j
            && self.smoke_source_kg == other.smoke_source_kg
            && self.heat_source_j == other.heat_source_j
            && self.carrier_boundary_kg == other.carrier_boundary_kg
            && self.smoke_boundary_kg == other.smoke_boundary_kg
            && self.heat_boundary_j == other.heat_boundary_j
    }
}
